src/Microsoft.AspNet.Server.Kestrel/Http/SocketInput.cs

```diff
@@ -35,80 +56,172 @@ public ArraySegment<byte> Take(int count)
-                    if (Buffer.Count != 0)
-                    {
-                        Buffer = new ArraySegment<byte>(Buffer.Array, 0, Buffer.Count);
-                    }
-                    availableSize = Buffer.Array.Length - Buffer.Offset - Buffer.Count;
+                    returnStart = _head;
+                    returnEnd = consumed.Block;
+                    _head = consumed.Block;
+                    _head.Start = consumed.Index;
                 }
-                else
+                if (!examined.IsDefault &&
+                    examined.IsEnd &&
+                    RemoteIntakeFin == false &&
+                    _awaitableError == null)
                 {
-                    var largerSize = Buffer.Array.Length + Math.Max(Buffer.Array.Length, minimumSize);
-                    var larger = new ArraySegment<byte>(_memory.AllocByte(largerSize), 0, Buffer.Count);
-                    if (Buffer.Count != 0)
-                    {
-                        Array.Copy(Buffer.Array, Buffer.Offset, larger.Array, 0, Buffer.Count);
-                    }
-                    _memory.FreeByte(Buffer.Array);
-                    Buffer = larger;
-                    availableSize = Buffer.Array.Length - Buffer.Offset - Buffer.Count;
+                    _manualResetEvent.Reset();
+
+                    var awaitableState = Interlocked.CompareExchange(
+                        ref _awaitableState,
+                        _awaitableIsNotCompleted,
+                        _awaitableIsCompleted);
                 }
             }
-            return new ArraySegment<byte>(Buffer.Array, Buffer.Offset + Buffer.Count, availableSize);
+            while (returnStart != returnEnd)
+            {
+                var returnBlock = returnStart;
+                returnStart = returnStart.Next;
+                returnBlock.Pool.Return(returnBlock);
+            }
+        }
+
+        public SocketInput GetAwaiter()
+        {
+            return this;
         }
 
-        public void Extend(int count)
+        public void OnCompleted(Action continuation)
         {
-            Debug.Assert(count >= 0);
-            Debug.Assert(Buffer.Offset >= 0);
-            Debug.Assert(Buffer.Offset <= Buffer.Array.Length);
-            Debug.Assert(Buffer.Offset + Buffer.Count <= Buffer.Array.Length);
-            Debug.Assert(Buffer.Offset + Buffer.Count + count <= Buffer.Array.Length);
+            var awaitableState = Interlocked.CompareExchange(
+                ref _awaitableState,
+                continuation,
+                _awaitableIsNotCompleted);
 
-            Buffer = new ArraySegment<byte>(Buffer.Array, Buffer.Offset, Buffer.Count + count);
+            if (awaitableState == _awaitableIsNotCompleted)
+            {
+                return;
+            }
+            else if (awaitableState == _awaitableIsCompleted)
+            {
+                Task.Run(continuation);
```

Comment:
Should we log any errors in the continuation in case the calling code doesn't properly handle it?